### PR TITLE
CW-997 Ip format validate 00 and 000

### PR DESCRIPTION
**Pull request by kuba2**

### CODE REVIEW

File: schema/formats_test.go [hunk 1]
```diff
@@ -90,6 +90,83 @@ var _ = Describe("format checkers", func() {
 			result := formatChecker.IsFormat("::1/129")
 			Expect(result).To(Equal(false))
 		})
+
+		It("Should not pass - two zeros in octet IPv4", func() {
```
Comment (zimnx) [inline schema/formats_test.go]: Change all of these tests to Table tests. ref: https://onsi.github.io/ginkgo/#table-driven-tests

File: schema/formats.go [hunk 2]
```diff
@@ -46,22 +47,53 @@ func (f macFormatChecker) IsFormat(input string) bool {
...
-	_, _, err := net.ParseCIDR(input)
-	return err == nil
+	ip, _, err := net.ParseCIDR(input)
+	if err != nil {
+		return false
+	}
+	if ip.To4() != nil {
+		return matchYangIPv4AddressPattern(input)
+	}
+	return true
```
Comment (zimnx) [inline schema/formats.go]: shouldn't it be false?

File: schema/formats.go [hunk 3]
```diff
@@ -46,22 +47,53 @@ func (f macFormatChecker) IsFormat(input string) bool {
 	return match
 }
 
+func matchYangIPv4AddressPattern(input string) bool {
+	res, err := regexp.MatchString(`^(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])(%[\p{N}\p{L}]+)?`, input)
```
Comment (zimnx) [inline schema/formats.go]: Have you handcrafted this regexp or took from some rfc?

File: schema/formats.go (hunk 2 above)
Comment (kuba2) [reply schema/formats.go]: This checker works for IPv6 and IPv4. Returning false would break IPv6 format checking.

File: schema/formats.go (hunk 3 above)
Comment (p-kozlowski) [inline schema/formats.go]: i suggest https://golang.org/pkg/regexp/#MustCompile for small efficiency gain

File: schema/formats.go [hunk 4]
```diff
@@ -46,22 +47,53 @@ func (f macFormatChecker) IsFormat(input string) bool {
...
+
+	if err != nil {
+		return false
+	}
+	return res
+}
+
+func (f ipv4FormatChecker) IsFormat(input string) bool {
+	ip := net.ParseIP(input)
+	if !((ip != nil) && (ip.To4() != nil)) {
```
Comment (p-kozlowski) [inline schema/formats.go]: for simplicity in code reading, I suggest helper functions + changing the `if` condition a little bit:

```Go
func isValid(ip net.IP) bool {
   return ip != nil
}

func isV4(ip net.IP) bool {
   return ip.To4() != nil
}

// in this function
if !isValid(ip) || !isV4(ip) {
   return false
}
```

File: schema/formats.go [hunk 5]
```diff
@@ -46,22 +47,53 @@ func (f macFormatChecker) IsFormat(input string) bool {
...
+	}
+	return true
 }
 
 func (f cidrOrIPv4FormatChecker) IsFormat(input string) bool {
 	cidrIP, _, cidrErr := net.ParseCIDR(input)
 	ip := net.ParseIP(input)
-	return (cidrErr == nil && cidrIP.To4() != nil) || (ip != nil && ip.To4() != nil)
+
+	if !((cidrErr == nil && cidrIP.To4() != nil) || (ip != nil && ip.To4() != nil)) {
```
Comment (p-kozlowski) [inline schema/formats.go]: again - can we simplify this condition?

File: schema/formats_test.go (hunk 1 above)
Comment (p-kozlowski) [reply schema/formats_test.go]: this file is a bad example, but prefer test names that read like sentences. now it's "CIDR format checker / Should not pass - two zeros in octet IPv4". I suggest "CIDR format checker / should reject two zeros in octet IPv4" instead

File: schema/formats_test.go [hunk 6]
```diff
@@ -90,6 +90,83 @@ var _ = Describe("format checkers", func() {
...
+			Expect(result).To(Equal(false))
+		})
+
+		It("Should not pass - three zeros in octet IPv4", func() {
+			result := formatChecker.IsFormat("127.000.0.1/24")
+			Expect(result).To(Equal(false))
+		})
+
+		It("Should not pass - two zeros in first octet IPv4", func() {
+			result := formatChecker.IsFormat("000.0.0.1/24")
```
Comment (p-kozlowski) [inline schema/formats_test.go]: test data does not match test name (should be "**00**.0...")

File: schema/formats.go [hunk 7]
```diff
@@ -46,22 +47,53 @@ func (f macFormatChecker) IsFormat(input string) bool {
...
+		return matchYangIPv4AddressPattern(input)
+	}
+	return true
 }
 
 func (f cidrOrIPv4FormatChecker) IsFormat(input string) bool {
 	cidrIP, _, cidrErr := net.ParseCIDR(input)
 	ip := net.ParseIP(input)
-	return (cidrErr == nil && cidrIP.To4() != nil) || (ip != nil && ip.To4() != nil)
+
```
Comment (p-kozlowski) [inline schema/formats.go]: could we reuse other format checkers here? sth. like

```Go
func (f cidrOrIPv4FormatChecker) IsFormat(input string) bool {
    ipv4 := ipv4FormatChecker{}
    cidr := ipv4NetworkFormatChecker{}
    return ipv4.IsFormat(input) || cidr.IsFormat(input)
}
```

File: schema/formats.go (hunk 3 above)
Comment (kuba2) [reply schema/formats.go]: Done

File: schema/formats.go (hunk 5 above)
Comment (kuba2) [reply schema/formats.go]: I reused format checkers and the condition is no longer here.

File: schema/formats_test.go (hunk 1 above)
Comment (kuba2) [reply schema/formats_test.go]: Good point. I changed descriptions to be more sentence like.

File: schema/formats.go (hunk 4 above)
Comment (kuba2) [reply schema/formats.go]: Good point. Looks simpler that way.

File: schema/formats.go (hunk 7 above)
Comment (kuba2) [reply schema/formats.go]: checkers are reused now.

File: schema/formats_test.go (hunk 6 above)
Comment (kuba2) [reply schema/formats_test.go]: done

File: schema/formats.go [hunk 8]
```diff
@@ -46,22 +47,53 @@ func (f macFormatChecker) IsFormat(input string) bool {
 	return match
 }
 
+func matchYangIPv4AddressPattern(input string) bool {
+	r := regexp.MustCompile(`^(([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])\.){3}([0-9]|[1-9][0-9]|1[0-9][0-9]|2[0-4][0-9]|25[0-5])(%[\p{N}\p{L}]+)?`)
```
Comment (p-kozlowski) [inline schema/formats.go]: this regexp will be compiled whenever a validator is invoked, possibly many times during single request processing. can we make it a package-level variable?

File: schema/formats_test.go [hunk 9]
```diff
@@ -51,158 +51,158 @@ var _ = Describe("format checkers", func() {
...
+			Entry("IPv4 cidr", "127.0.0.1/16"),
+			Entry("IPv4 network", "127.0.0.0/16"),
+			Entry("IPv6 cidr", "::1/16"),
+			Entry("IPv6 network", "fe80::9e17/64"),
+			Entry("wrong IPv4", "218.108.149.379"),
+			Entry("wrong IPv4 with three octets only", "218.149.179"),
+			Entry("IPv4 with two zeros in octet ", "127.00.0.1"),
+			Entry("IPv4 with three zeros in octet ", "127.000.0.1"),
+			Entry("IPv4 with two zeros in the first octet ", "00.3.0.1"),
+			Entry("IPv4 wiht three zeros in the first octet ", "000.4.0.1"),
```
[the remark this answers is not in the document]
Comment (kuba2) [reply schema/formats_test.go]: fixed

File: schema/formats.go (hunk 8 above)
Comment (kuba2) [reply schema/formats.go]: made a package level variable

File: schema/formats.go [hunk 10]
```diff
@@ -41,27 +42,59 @@ type textFormatChecker struct{}
...
+}
+
 func (f cidrFormatChecker) IsFormat(input string) bool {
-	_, _, err := net.ParseCIDR(input)
-	return err == nil
+	ip, _, err := net.ParseCIDR(input)
+	if err != nil {
+		return false
+	}
+	if isV4(ip) {
```
Comment (jkmar) [inline schema/formats.go]: there is also yang for ipv6:
https://tools.ietf.org/html/rfc6991#page-19

**Pull request merged**